added batch logging

AceTheNinja opened the pull request.

AceTheNinja committed:
added batch logging

btrdnch commented on index.js:
```diff
@@ -9,9 +9,39 @@ class GrcSlack extends Base {
+      this._errorBatchTimer = null
+    }
+
+    this._processBatchedErrors().catch(err => {
```
please change linting here

btrdnch commented on README.md:
```diff
@@ -11,7 +11,12 @@ A facility that simplifies the integration of slack message service  via greanch
+    "env": "development",
+    "errorBatching": { // optional
+      "interval": 60000,
+      "maxSize": 50,
```
is it possible to compose and send 50x4000 symbols to slack at once?

AceTheNinja replied:
We can create and send a file in that case

btrdnch replied:
no files

btrdnch commented on index.js:
```diff
@@ -41,6 +71,137 @@ class GrcSlack extends Base {
+   * @param {Object} payload - Payload to log
+   * @param {...any} extra - Additional information to log
+   */
+  async batchLogErrorToSlack (reqChannel, err, functionName, payload, ...extra) {
```
```suggestion
  async logErrorEnqueue (reqChannel, err, ...extra) {
```

btrdnch commented on index.js:
```diff
@@ -41,6 +71,137 @@ class GrcSlack extends Base {
+      errorEntry.lastSeen = now
+      errorEntry.payloads.push(payload)
+
+      if (this._errorBatch.size >= this._errorBatchingConfig.maxSize) {
```
i don't think we should flush if buffer is full. We should evict oldest keys. Maybe it would be better to use lru here

btrdnch commented on index.js:
```diff
@@ -41,6 +71,137 @@ class GrcSlack extends Base {
+    message += `*Summary:* ${totalErrors} errors across ${errors.length} types (${timeRange})\n\n`
+
+    let truncated = false
+    for (const error of errors.slice(0, 10)) {
```
why do we slice?

AceTheNinja replied:
If the message is too large the API will fail so have kept it to only 10 different error per function

btrdnch replied:
what happens to the rest?

btrdnch replied:
question still stands. I suggest we log only last to slack to fulfill length requirement

btrdnch commented on index.js:
```diff
@@ -41,6 +71,137 @@ class GrcSlack extends Base {
+    for (const error of errors.slice(0, 10)) {
+      if (truncated) break
+
+      message += `• *${error.errorMessage}* (${error.count}x)\n`
```
what about the rest message info? like extra, etc.

AceTheNinja replied:
that goes in payload

AceTheNinja committed:
added lru based implementation

btrdnch commented on index.js:
```diff
@@ -12,6 +13,41 @@ class GrcSlack extends Base {
     if (opts.conf) this.conf = opts.conf
+
+    this._errorBatchingConfig = {
+      interval: this.conf.errorBatching.interval || 60000,
```
let's assume `errorBatching` may be absent completely cause we're not going to update all existing running places

AceTheNinja replied:
added null coalesce

btrdnch commented on index.js:
```diff
@@ -41,6 +77,137 @@ class GrcSlack extends Base {
+   * @param {Object} payload - Payload to log
+   * @param {...any} extra - Additional information to log
+   */
+  async logErrorEnqueue (reqChannel, err, functionName, payload, ...extra) {
```
let's make parameter `functionName` more generalized, i.e. `errorOrigin` or `originName` or `errorSource` or `sourceName` etc. You can come up with something better.

btrdnch commented on index.js:
```diff
@@ -41,6 +77,137 @@ class GrcSlack extends Base {
+   * @param {Error} err - Error to log
+   * @param {string} functionName - Name of the function where the error occurred
+   * @param {Object} payload - Payload to log
+   * @param {...any} extra - Additional information to log
```
is extra a required parameter?

AceTheNinja replied:
no we're just putting all other params in 1 object. It can be undefined also

btrdnch replied:
```suggestion
   * @param {...any} [extra] - Additional information to log
```

btrdnch commented on index.js:
```diff
@@ -41,6 +77,137 @@ class GrcSlack extends Base {
+      let errorEntry = this._errorBatch.get(errorKey)
+
+      if (!errorEntry) {
+        errorEntry = {
```
`extra` is not used, what is it for then?

AceTheNinja replied:
added it

AceTheNinja committed:
functionName param change

AceTheNinja committed:
added extra to logs

AceTheNinja committed:
removed unnecessary comments

btrdnch commented on index.js:
```diff
@@ -12,6 +13,41 @@ class GrcSlack extends Base {
 
     if (opts.conf) this.conf = opts.conf
+
+    this._errorBatchingConfig = {
```
why do we store it again if we already have it in `this.conf`?
additionally, if it's omitted (in readme it's commented as optional), you gonna get error trying to access `this.conf.errorBatching`
if there is no `this.conf.errorBatching`, this instance is not using batching, we should not setup it. I think additional timer is an overhead cause not all workers need batching

AceTheNinja replied:
removed extra variable and not setting up batch logging interval if config does not exist

btrdnch commented on index.js:
```diff
@@ -12,6 +13,41 @@ class GrcSlack extends Base {
+      maxMessageLength: this.conf.errorBatching?.maxMessageLength || 4000
+    }
+
+    this._errorBatch = new LRU({
```
if there is no `this.conf.errorBatching`, we should not setup it

btrdnch commented on index.js:
```diff
@@ -12,6 +13,41 @@ class GrcSlack extends Base {
+      max: this._errorBatchingConfig.maxSize
+    })
+
+    this._initErrorBatching()
```
if there is no `this.conf.errorBatching`, we should not setup it

btrdnch commented on test/facility.test.js:
pls create folder `test` and put a file there named `facility.test.js`

AceTheNinja replied:
done

btrdnch commented on package.json:
```diff
@@ -13,9 +13,12 @@
+    "lru": "^3.1.0"
   },
   "devDependencies": {
+    "@types/jest": "^30.0.0",
```
you need it?

AceTheNinja replied:
needed it for local syntax highlighting. Removed it

btrdnch commented on package.json:
```diff
@@ -13,9 +13,12 @@
   "dependencies": {
-    "bfx-facs-base": "git+https://github.com/bitfinexcom/bfx-facs-base.git"
+    "bfx-facs-base": "git+https://github.com/bitfinexcom/bfx-facs-base.git",
+    "lru": "^3.1.0"
```
let's lock dep and add package-lock.json (use node 16.20)

AceTheNinja replied:
added

btrdnch commented on index.js:
```diff
@@ -41,6 +77,140 @@ class GrcSlack extends Base {
+    const earliest = new Date(Math.min(...allTimes))
+    const latest = new Date(Math.max(...allTimes))
+
+    const formatTime = (date) => date.toISOString().substring(11, 19)
```
looks like a utility function that should not be re-defined inside a function each time

AceTheNinja replied:
moved to utils

btrdnch commented on index.js:
```diff
@@ -41,6 +77,140 @@ class GrcSlack extends Base {
+        await this._sendBatchedErrorMessage(reqChannel, sourceName, errors)
+      }
+
+      this._errorBatch.clear()
```
if `_sendBatchedErrorMessage` fails, we never clear?

AceTheNinja replied:
now it should, added clear to finally block

AceTheNinja committed:
- improved conf handling
- batching init is now conf based
- moved test to test/facility.test.js
- clear batch if sending fails so error don't pile up

AceTheNinja committed:
added package-lock with node 16

AceTheNinja committed:
removed @types/jest from devDependencies

AceTheNinja commented:
https://github.com/bitfinexcom/bfx-facs-grc-slack/pull/5#discussion_r2369283829

cannot have optional parameter with spread operator. If no args are present it will be empty

avsek477 commented on package.json:
```diff
@@ -13,10 +13,12 @@
   "dependencies": {
-    "bfx-facs-base": "git+https://github.com/bitfinexcom/bfx-facs-base.git"
+    "bfx-facs-base": "git+https://github.com/bitfinexcom/bfx-facs-base.git",
+    "lru": "3.1.0"
```
why not use facs-lru ?

AceTheNinja replied:
adding another fac that implements base and starting it in the constructor felt like an unnecessary complication for something simple like this

btrdnch replied:
@AceTheNinja let's add an optional lru fac dependency for slack for the cases when we enable batch logging

AceTheNinja committed:
refactor: add lru and optional argument to fac

vigan-abd commented on index.js:
```diff
@@ -12,6 +14,34 @@ class GrcSlack extends Base {
+    }
+
+    try {
+      await this._processBatchedErrors()
```
why this is called on stop? wouldn't it hang for a long time possibly?

btrdnch replied:
i think it's called to flush the last batch of errors, otherwise it'll be lost
not sure about long time, i think it should be reasonable time since it's an api call. We can limit the call / stop function in time and exit abruptly if we exceed graceful exit limit.

vigan-abd commented on utils/date-time.js:
use strict missing

AceTheNinja replied:
pushed an update

vigan-abd commented on index.js:
```diff
@@ -12,6 +14,34 @@ class GrcSlack extends Base {
 
     if (opts.conf) this.conf = opts.conf
+
+    if (this.conf.errorBatching && opts.lru) {
```
why are we using lru and not map? with lru we would lose data right?

btrdnch replied:
i think in order to keep the buffer fixed in length. If a worker starts to spam, we don't want to keep the spam rather than track let's say last 100 messages of a certain type.
for instance, a worker sends error messages for different occasions, i.e. API call failure and processing issue. Let's say it reports 1 api call failure and 1000 processing issues, We want both types to be reported, however we don't want 1000 processing issues (same type issue) being reported, we cut it down to, let's say, 10 only

that was the idea behind this update in general. If a worker starts spamming, this logic would still allow us to know about the problem, however, would cut down api calls and payload significantly.

anyways, if i missed the purppose, @AceTheNinja please clarify

AceTheNinja replied:
couldn't have explained it better 🫡  @batrudinych

AceTheNinja committed:
chore: enable strict mode in date-time utility

btrdnch commented on index.js:
```diff
@@ -21,7 +51,7 @@ class GrcSlack extends Base {
     const env = (slack.env) ? `Env: ${slack.env}, ` : ''
     const rawText = env + message
-    const text = (maxLength) ? rawText.substr(0, maxLength) : rawText
+    const text = (maxLength) ? rawText.slice(0, maxLength) : rawText
```
```suggestion
    const text = (rawText.length > maxLength) ? rawText.slice(0, maxLength) : rawText
```

btrdnch commented on index.js:
```diff
@@ -41,6 +71,143 @@ class GrcSlack extends Base {
+  }
+
+  _createErrorKey (reqChannel, err, sourceName) {
+    const errorMsg = err?.message || 'Unknown error'
```
[check this](https://github.com/bitfinexcom/bfx-facs-grc-slack/blob/e4cc25858633bb47230aab425cf79c0a9cc7b807/index.js#L66)

maybe you compute hash of stringified error instead of using it directly?

btrdnch commented on index.js:
```diff
@@ -41,6 +71,143 @@ class GrcSlack extends Base {
+
+      errorEntry.count++
+      errorEntry.lastSeen = now
+      errorEntry.payloads.push({ payload, extras: extra })
```
you're adding same stuff twice if you just initialized cache entry.

btw if we spam under same errorKey, aren't we bloating the payloads array? Size should be restricted, otherwise we can run out of memory

btrdnch commented on index.js:
```diff
@@ -41,6 +71,143 @@ class GrcSlack extends Base {
+          payloads: [
+            { payload, extras: extra }
+          ],
+          count: 0,
```
is it 0? you just seen it first time

btrdnch commented on index.js:
```diff
@@ -41,6 +71,143 @@ class GrcSlack extends Base {
+    }
+  }
+
+  _createErrorKey (reqChannel, err, sourceName) {
```
make a default value for sourceName

btrdnch commented on index.js:
```diff
@@ -41,6 +71,143 @@ class GrcSlack extends Base {
+
+      const allEntries = Object.values(this._errorBatch.cache.cache || {})
+
+      for (const { value: errorEntry } of allEntries) {
```
the part below is overcomplicated. You already grouped by channel + message + source. Each grouped item has a list of payloads. Go over each item, concatenate payloads until you reach restricted length and send a message, continue until the lru is exhausted. Pay attention to not hit limits while you're doing so (i.e. overall  messages stored should not be too high relatively to the length of a single message sent so we wouldn't send too many msgs too quickly)

btrdnch commented on index.js:
```diff
@@ -41,6 +71,143 @@ class GrcSlack extends Base {
+      }
+
+      for (const [groupKey, errors] of errorsByFunctionNameAndChannel) {
+        const [reqChannel, sourceName] = groupKey.split(':')
```
not good. We compose keys 3 times to this point, can be avoided.

additionally, if you somehow get too many errors under same key, going to spam the api

AceTheNinja replied:
Moved the key calculation to a function and removed from one place. grouping is important so cannot skip this logic.

We're only keeping the last 3 payloads and flush happens at intervals so no spamming

btrdnch commented on index.js:
```diff
@@ -41,6 +71,143 @@ class GrcSlack extends Base {
+      message += `• *${error.errorMessage}* (${error.count}x)\n`
+      message += '  Payloads:\n'
+
+      for (const item of error.payloads.slice(0, 3)) {
```
let's take latest and make this configurable, last X

btrdnch commented on index.js:
```diff
@@ -41,6 +71,143 @@ class GrcSlack extends Base {
+          payloadStr += `     Extras: ${JSON.stringify(item.extras)}\n`
+        }
+
+        if (message.length + payloadStr.length > (this.conf.errorBatching?.maxMessageLength || 4000)) {
```
set default once when you initiate, not here

btrdnch commented on index.js:
```diff
@@ -41,6 +71,143 @@ class GrcSlack extends Base {
+        message += payloadStr
+      }
+
+      if (error.payloads.length > 3) {
```
pls no magic numbers

btrdnch commented on index.js:
```diff
@@ -41,6 +71,143 @@ class GrcSlack extends Base {
+      }
+    }
+
+    if (errors.length > 10 && !truncated) {
```
no magic numbers

AceTheNinja replied:
removed

btrdnch commented on index.js:
```diff
@@ -41,6 +71,143 @@ class GrcSlack extends Base {
+  }
+
+  async _sendBatchedErrorMessage (reqChannel, sourceName, errors) {
+    const totalErrors = errors.reduce((sum, error) => sum + error.count, 0)
```
you can keep the counter separately to avoid computing it here, 1 less array pass

btrdnch commented on index.js:
```diff
@@ -41,6 +71,143 @@ class GrcSlack extends Base {
+            { payload, extras: extra }
+          ],
+          count: 0,
+          firstSeen: now,
```
looks like you use these dates just to know when you got first and last errors. Use separate properties, do not store inside the object and avoid excessive array passes

btrdnch commented on package.json:
```diff
@@ -15,8 +15,12 @@
     "bfx-facs-base": "git+https://github.com/bitfinexcom/bfx-facs-base.git"
   },
+  "optionalDependencies": {
+    "bfx-facs-lru": "git+https://github.com/bitfinexcom/bfx-facs-lru.git"
```
do we need it here? it should come from outside as dependency when we instantiate the fac, i think

AceTheNinja replied:
yeah not needed

AceTheNinja committed:
Update index.js

Co-authored-by: batrudinych <batrudinych@gmail.com>

AceTheNinja committed:
refactor: update error batching logic and key generation

AceTheNinja committed:
Merge branch 'main' of github.com:AceTheNinja/bfx-facs-grc-slack

AceTheNinja committed:
removed array passes

AceTheNinja committed:
refactor: added error batching config validation and improve message truncation logic

AceTheNinja committed:
removed optional deps
